File: solutions/cpp/Computing_binomial_coefficients.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <random>
#include <vector>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::min;
using std::random_device;
using std::uniform_int_distribution;
using std::vector;

int ComputeXChooseY(int, int, vector<vector<int>>*);

// @include
int ComputeBinomialCoefficient(int n, int k) {
  vector<vector<int>> x_choose_y(n + 1, vector<int>(k + 1, 0));
  return ComputeXChooseY(n, k, &x_choose_y);
}

int ComputeXChooseY(int x, int y, vector<vector<int>>* x_choose_y_ptr) {
  if (y == 0 || x == y) {
    return 1;
  }

  vector<vector<int>>& x_choose_y = *x_choose_y_ptr;
  if (x_choose_y[x][y] == 0) {
    int without_y = ComputeXChooseY(x - 1, y, x_choose_y_ptr);
    int with_y = ComputeXChooseY(x - 1, y - 1, x_choose_y_ptr);
    x_choose_y[x][y] = without_y + with_y;
  }
  return x_choose_y[x][y];
}
// @exclude

int ComputeBinomialCoefficientsSpaceEfficient(int n, int k) {
  k = min(k, n - k);
  vector<int> table(k + 1, 0);
  table[0] = 1;  // C(0, 0).
  // C(i, j) = C(i - 1, j) + C(i - 1, j - 1).
  for (int i = 1; i <= n; ++i) {
    for (int j = min(i, k); j >= 1; --j) {
      table[j] = table[j] + table[j - 1];
    }
    table[0] = 1;  // One way to select zero element.
  }
  return table[k];
}

int CheckAns(int n, int k) {
  vector<int> number;
  for (int i = 0; i < k; ++i) {
    number.emplace_back(n - i);
  }

  vector<int> temp;
  for (int i = 2; i <= k; ++i) {
    bool find = false;
    for (int& a : number) {
      if ((a % i) == 0) {
        a /= i;
        find = true;
        break;
      }
    }
    if (find == false) {
      temp.emplace_back(i);
    }
  }

  int res = 1;
  for (int a : number) {
    res *= a;
  }

  for (int a : temp) {
    res /= a;
  }

  return res;
}

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int n, k;
    if (argc == 3) {
      n = atoi(argv[1]), k = atoi(argv[2]);
    } else {
      uniform_int_distribution<int> n_dis(1, 21);
      n = n_dis(gen);
      uniform_int_distribution<int> k_dis(0, n);
      k = k_dis(gen);
    }

    int res = ComputeBinomialCoefficient(n, k);
    cout << "res = " << res << endl;
    assert(res == CheckAns(n, k));
    assert(res == ComputeBinomialCoefficientsSpaceEfficient(n, k));
    cout << n << " out of " << k << " = " << res << endl;
    if (argc == 3) {
      break;
    }
  }
  return 0;
}

File: static/python/computing_binomial_coefficients.py
import sys
import random


# @include
def compute_binomial_coefficient(n, k):
    def compute_x_choose_y(x, y):
        if y in (0, x):
            return 1

        if x_choose_y[x][y] == 0:
            without_y = compute_x_choose_y(x - 1, y)
            with_y = compute_x_choose_y(x - 1, y - 1)
            x_choose_y[x][y] = without_y + with_y
        return x_choose_y[x][y]

    x_choose_y = [[0] * (k + 1) for _ in range(n + 1)]
    return compute_x_choose_y(n, k)
# @exclude


def compute_binomial_coefficients_space_efficient(n, k):
    k = min(k, n - k)
    table = [1] + [0] * k
    # C(i, j) = C(i - 1, j) + C(i - 1, j - 1).
    for i in range(1, n + 1):
        for j in reversed(range(1, min(i, k) + 1)):
            table[j] = table[j] + table[j - 1]
    return table[-1]


def check_ans(n, k):
    number = [n - i for i in range(k)]
    temp = []
    for i in range(2, k + 1):
        find = False
        for a in number:
            if a % i == 0:
                a //= i
                find = True
                break
        if not find:
            temp.append(i)

    res = 1
    for a in number:
        res *= a

    for a in temp:
        res //= a

    return res


def main():
    for times in range(1000):
        if len(sys.argv) == 3:
            n = int(sys.argv[1])
            k = int(sys.argv[2])
        else:
            n = random.randint(1, 21)
            k = random.randint(0, n)

        res = compute_binomial_coefficient(n, k)
        print('res =', res)
        assert res == compute_binomial_coefficients_space_efficient(n, k)
        print(n, 'out of', k, '=', res)
        if len(sys.argv) == 3:
            break


if __name__ == '__main__':
    main()
